### DISCUSSION TITLE
Collections empty after beatmap realm migration

### DISCUSSION THREAD

**Discussion by zachmanthethird** [ppy/osu#16517] [Q&A] [answered]
I hate opening a discussion without proof, but I didn't have the foresight to back up my client.realm file for replication purposes.
I also don't have any logs from the migration process as I didn't catch it fast enough...
Version local debug; latest commit 5809ce69513f075a1b039a50cfef4ef9e0185312
I know this makes things difficult to replicate. Feel free to just say this isn't enough information.
[collection.db.txt](https://github.com/ppy/osu/files/7893871/collection.db.txt)

**Comment by peppy** [comment]
There isn't any migration of collections with the realm change. Can you confirm what you are currently seeing? Your collections are showing but there are no beatmaps in them?

Could you also provide your realm file so I can check why they aren't mapping correctly?

**Reply by zachmanthethird** [reply]
Correct. The most I can confirm is that the collections were just fine before the migration, and they are now empty.
I attempted to run version 2022.118.0; however, this obviously failed due to different realm versions.
It doesn't make logical sense that the two are related; I agree.

**Reply by peppy** [reply]
I still need your `client.realm` file if possible. You can zip and attach it here.

**Reply by smoogipoo** [reply]
The `collection.db` too, would also be helpful.

**Reply by zachmanthethird** [reply]
[client.realm.txt](https://github.com/ppy/osu/files/7894012/client.realm.txt)
Again, I wish I had backed up my copy before the migration; I will be sure to do so in the future.

**Reply by zachmanthethird** [reply]
Luckily, my collections correspond to tournament mappools, so recreating them wasn't a problem. This should be what the collections looked like before: [collection.db.txt](https://github.com/ppy/osu/files/7894082/collection.db.txt)

**Reply by zachmanthethird** [reply]
> I attempted to run version 2022.118.0; however, this obviously failed due to different realm versions.

So the closest I can replicate is by removing/renaming the client.realm file causes the collection.db to empty of beatmaps. I don't remember doing this specifically, but there may be a similar case where a different version of realm does the same thing.
This may be intended behavior and just needing some form of warning/error to be logged/shown.
[client.realm.txt](https://github.com/ppy/osu/files/7894330/client.realm.txt) generated due to the absence of one.
[database.log.txt](https://github.com/ppy/osu/files/7894332/database.log.txt) [network.log.txt](https://github.com/ppy/osu/files/7894333/network.log.txt) [runtime.log.txt](https://github.com/ppy/osu/files/7894334/runtime.log.txt)

**Reply by peppy** [reply]
~~I don't suppose you have your original `client.db` as well? If you migrated after the backup PR was merged there should be a renamed copy of it.~~ really can't investigate this further without a copy of the original `collections.db` as well.

**Reply by peppy** [reply]
I've attempted to repro the issue you had locally from scratch and can't. Which is a touch worrying.

The only thing I can think of is that we need to force a `Realm.Refresh` after the migration completes. It would run anyway on the first update call of the game, and `CollectionManager` is chain-loaded post that point in time (in theory?) so I doubt this could be the issue, but I can't see any other potential reasons for this failing.

Are you sure you didn't migrate on any other commit? There may have been some bugs in `master` which caused incorrect migrations at certain points.

**Comment by zachmanthethird** [comment]
> Are you sure you didn't migrate on any other commit? There may have been some bugs in `master` which caused incorrect migrations at certain points.

So I've updated to latest commit of 9920ff51a62bfa080e7f5c9a020136658661d010 and it definitely looks like this is true.
[runtime.log.txt](https://github.com/ppy/osu/files/7894961/runtime.log.txt)

> So the closest I can replicate is by removing/renaming the client.realm file causes the collection.db to empty of beatmaps.
This may be intended behavior and just needing some form of warning/error to be logged/shown.

This appears to be consistent. While I believe my exact situation is unique, I believe this general outcome is possible due to failed migrations via #16495 
This looks like it'll be fixed via #16521; will report once I can confirm.

**Reply by zachmanthethird** [reply]
If there is no client.realm file, the collection.db is still written without a backup. If you think this isn't a possible situation to end up in, please feel free to mark the above as the answer.

**Reply by peppy** [reply]
That scenario can't happen with realm migration. It would crash the game before that happened.

**Comment by peppy** [comment, marked as answer]
"Resolved" via https://github.com/ppy/osu/pull/16521#event-5915724066
